pettingzoo/mpe/scenarios/group_spread.py

```diff
@@ -150,34 +142,26 @@ def global_reward(self, world):
         return rew
 
     def observation(self, agent, world):
+        # own position, velocity, color
+        color = [np.eye(self.num_colors)[agent.color_idx]] if self.obs_onehot_colors else [agent.color]
+        entity_features = [agent.state.p_pos] + [agent.state.p_vel] + color
+
         # positions and color of all landmarks
-        entity_pos_color = []
-        if self.shuffle_obs_per_agent:
-            landmark_idxs = agent.landmark_obs_order
-        else:
-            landmark_idxs = np.arange(len(world.landmarks))
-        for i in landmark_idxs:
-            entity_pos_color.append(world.landmarks[i].state.p_pos - agent.state.p_pos)
-            if self.randomise_all_colors:
-                entity_pos_color.append(world.landmarks[i].color)
-            else:
-                # one hot encoding
-                entity_pos_color.append(np.eye(self.num_colors)[world.landmarks[i].color_idx])
```

Comment:
This code conflates one-hot encoding of colors and `randomise_all_colors=True` we use real-valued colors, if `randomise_all_colors=False` then we use one-hot colors for landmarks.

This is imo quite unintuitive as the flag does not suggest this is happening and I also don't see why we would conflate those two effects in this way (isn't it reasonable to use one-hot colors in agent observations even if we use randomised colors, and also to use real-valued colors in observations even if we use consistent colors).

To address this, I use a new `obs_onehot_colors` flag which is true by default and can determine whether all color values in observations (own, other agents, landmarks) are given as onehot (True; default) or real-valued arrays (False).

Reply:
I'm fine with a new flag here, I think it makes it clearer. However, I would add the one-hot encoding when randomising all colours doesn't make sense - that would be a very big vector to allow for all colours! So maybe we want to catch that and raise an exception.